crates/runtime/src/table.rs

```diff
@@ -328,10 +460,46 @@ impl Table {
         Ok(Some(old_size))
     }
 
+    /// Returns the lazy-init FuncIndex table, if any.
+    fn lazy_data(&self) -> Option<&Arc<TableLazyData>> {
+        match self {
+            Table::Static { lazy_data, .. } | Table::Dynamic { lazy_data, .. } => {
+                lazy_data.as_ref()
+            }
+        }
+    }
+
     /// Get reference to the specified element.
     ///
     /// Returns `None` if the index is out of bounds.
-    pub fn get(&self, index: u32) -> Option<TableElement> {
+    pub fn get_or_init(
+        &mut self,
+        index: u32,
+        mut instance: InstanceHandle,
```

Comment:
In general `InstanceHandle` is intended to roughly represent an owned instance, so I don't think that it's the best choice for this type parameter.

I think that at the `Table` layer here it's best for `get` to return `enum { None, NotInitialized, Some(element) }` and for an upper layer to handle the lazy initialization. Notably I think that `Instance` would be a good place to handle the "do the get, if that fails, do a set" since that's where the data for the lazily initialized slot will be stored anyway.

Reply:
Reworked this bit along the lines you suggest:

- the `TableElement` has an `UninitFunc` arm;
- There is a separate `TableAndOwningInstance` handle-like struct that encapsulates a `Table` and an `Instance` and knows how to do lazy init (and can hand out projected borrows to the underlying table);
- We get this `TableAndOwningInstance` in the various spots where a table is looked up on an instance.

The design is still ever so slightly awkward (see comment on `InstanceHandle::from_instance_mut`) because it needs to be available both at the libcall layer and in the `wasmtime::Table` implementation, but this is a lot cleaner for sure.